libs/libapi/src/libapi/duckdb.py

```diff
@@ -0,0 +1,119 @@
+        split_directory = url.rsplit("/", 2)[1]
+        repo_file_location = f"{config}/{split_directory}/{filename}"
+        index_file_location = f"{index_folder}/{repo_file_location}"
+        index_path = Path(index_file_location)
+        if not index_path.is_file():
+            with StepProfiler(method="get_index_file_location_and_download_if_missing", step="download index file"):
+                download_index_file(
+                    cache_folder=f"{duckdb_index_file_directory}/{HUB_DOWNLOAD_CACHE_FOLDER}",
+                    index_folder=index_folder,
+                    target_revision=target_revision,
+                    dataset=dataset,
+                    repo_file_location=repo_file_location,
+                    hf_token=hf_token,
+                )
+        return index_file_location
+
+
+def get_download_folder(
+    root_directory: StrPath, dataset: str, config: str, split: str, revision: Optional[str]
+) -> str:
+    payload = (dataset, config, split, revision)
+    hash_suffix = sha1(json.dumps(payload, sort_keys=True).encode(), usedforsecurity=False).hexdigest()[:8]
+    subdirectory = "".join([c if re.match(r"[\w-]", c) else "-" for c in f"{dataset}-{hash_suffix}"])
+    return f"{root_directory}/downloads/{subdirectory}"
+
+
+def download_index_file(
+    cache_folder: str,
+    index_folder: str,
+    target_revision: str,
+    dataset: str,
+    repo_file_location: str,
+    hf_token: Optional[str] = None,
+) -> None:
+    logging.info(f"init_dir {index_folder}")
+    init_dir(index_folder)
+
+    # see https://pypi.org/project/hf-transfer/ for more details about how to enable hf_transfer
+    os.environ["HF_HUB_ENABLE_HF_TRANSFER"] = "1"
+    hf_hub_download(
+        repo_type=REPO_TYPE,
+        revision=target_revision,
+        repo_id=dataset,
+        filename=repo_file_location,
+        local_dir=index_folder,
+        local_dir_use_symlinks=False,
+        token=hf_token,
+        cache_dir=cache_folder,
+    )
+
+
+def get_cache_entry_from_duckdb_index_job(
+    processing_graph: ProcessingGraph,
+    dataset: str,
+    config: str,
+    split: str,
+    hf_endpoint: str,
+    hf_token: Optional[str],
+    hf_timeout_seconds: Optional[float],
+    cache_max_days: int,
+    blocked_datasets: list[str],
+) -> CacheEntry:
+    processing_steps = processing_graph.get_processing_step_by_job_type("split-duckdb-index")
+    return get_cache_entry_from_steps(
+        processing_steps=[processing_steps],
+        dataset=dataset,
+        config=config,
+        split=split,
+        processing_graph=processing_graph,
+        hf_endpoint=hf_endpoint,
+        hf_token=hf_token,
+        hf_timeout_seconds=hf_timeout_seconds,
+        cache_max_days=cache_max_days,
+        blocked_datasets=blocked_datasets,
+    )
+
+
+def duckdb_connect(**kwargs: Any) -> duckdb.DuckDBPyConnection:
+    con = duckdb.connect(read_only=True, **kwargs)
+    con.sql("SET enable_external_access=false;")
```

Comment:
we never use duckdb to directly access the remote parquet files or duckdb index file stored in refs/convert/parquet, right?

Reply:
https://github.com/huggingface/datasets-server/pull/1418#issuecomment-1742788076
> EDIT: setting enable_external_access to false seems to block http requests and globbing local files!

Reply:
@severo I would say that this duckdb connection is only used with the downloaded local index because the `duckdb_connect` is only used in the /search and /filter endpoints, both with the argument `index_file_location`, that is the path of the previously downloaded index duckdb file (which we download with `get_index_file_location_and_download_if_missing`).

Therefore I'm merging this PR; anyway we could revert this configuration setting if necessary before deploying.